[ESPHome 2024.12.x] ESPNow won't compile with version 2024.12.x of ESPHome

### The problem

```
Compiling .pioenvs/x83fwa/src/esphome/components/logger/logger_esp32.cpp.o
src/esphome/components/esp_now/esp_now_component.cpp: In member function 'virtual void esphome::esp_now::ESPNowComponent::setup()':
src/esphome/components/esp_now/esp_now_component.cpp:32:33: error: invalid conversion from 'void (*)(const uint8_t*, const uint8_t*, int)' {aka 'void (*)(const unsigned char*, const unsigned char*, int)'} to 'esp_now_recv_cb_t' {aka 'void (*)(const esp_now_recv_info*, const unsigned char*, int)'} [-fpermissive]
   32 |   err = esp_now_register_recv_cb(ESPNowComponent::on_data_received);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                 |
      |                                 void (*)(const uint8_t*, const uint8_t*, int) {aka void (*)(const unsigned char*, const unsigned char*, int)}
In file included from src/esphome/components/esp_now/esp_now_component.cpp:4:
/data/cache/platformio/packages/framework-espidf/components/esp_wifi/include/esp_now.h:157:54: note:   initializing argument 1 of 'esp_err_t esp_now_register_recv_cb(esp_now_recv_cb_t)'
  157 | esp_err_t esp_now_register_recv_cb(esp_now_recv_cb_t cb);
      |                                    ~~~~~~~~~~~~~~~~~~^~
*** [.pioenvs/x83fwa/src/esphome/components/esp_now/esp_now_component.cpp.o] Error 1
```

### Which version of ESPHome has the issue?

2024.12.0b1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

_No response_

### Component causing the issue

_No response_

### YAML Config

```yaml
substitutions:
  id: x83fwa
  model: NodeMCU 32
  name: Tracker
  ip: 192.168.100.41
esphome:
  name: x83fwa
  comment: NodeMCU 32
  friendly_name: Tracker
  min_version: 2024.12.0
  build_path: build/x83fwa
  area: ''
  platformio_options: {}
  includes: []
  libraries: []
  name_add_mac_suffix: false
esp32:
  board: nodemcu-32s
  framework:
    version: 5.1.5
    sdkconfig_options: {}
    advanced:
      ignore_efuse_custom_mac: false
    components: []
    platform_version: https://github.com/pioarduino/platform-espressif32.git#51.03.07
    source: pioarduino/framework-espidf@https://github.com/pioarduino/esp-idf/releases/download/v5.1.5/esp-idf-v5.1.5.zip
    type: esp-idf
  flash_size: 4MB
  variant: ESP32
wifi:
  use_address: 192.168.2.41
  ap:
    ssid: ESPHome Fallback
    password: xxx
    ap_timeout: 1min
  domain: .local
  reboot_timeout: 15min
  power_save_mode: LIGHT
  fast_connect: false
  enable_btm: false
  enable_rrm: false
  passive_scan: false
  enable_on_boot: true
  networks:
  - ssid: xxx
    password: xxx
    priority: 0.0
logger:
  baud_rate: 115200
  tx_buffer_size: 512
  deassert_rts_dtr: false
  hardware_uart: UART0
  level: DEBUG
  logs: {}
api:
  encryption:
    key: xxx
  port: 6053
  password: ''
  reboot_timeout: 15min
ota:
- platform: esphome
  password: xxx
  version: 2
  port: 3232
web_server:
  port: 80
  version: 2
  enable_private_network_access: true
  include_internal: false
  ota: false
  log: true
  css_url: ''
  js_url: https://oi.esphome.io/v2/www.js
captive_portal: {}
time:
- platform: homeassistant
  id: homeassistant_time
  timezone: CET-1CEST,M3.5.0,M10.5.0/3
  update_interval: 15min
external_components:
- source:
    url: https://github.com/jesserockz/wizmote-esphome.git
    type: git
  components:
  - esp_now
  - wizmote
  refresh: 1d
esp_now: {}
bluetooth_proxy:
  active: true
  cache_services: true
  connections:
  - {}
  - {}
  - {}
wizmote:
  on_button:
    then:
    - homeassistant.event:
        event: esphome.wizmote_action
        data:
          mac: !lambda |-
            return format_hex(data.bssid, 6);
          button: !lambda |-
            return data.button;
          sequence: !lambda |-
            return data.sequence;
        data_template: {}
        variables: {}
esp32_ble_tracker:
  scan_parameters:
    duration: 5min
    interval: 320ms
    window: 30ms
    active: true
    continuous: true
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

Linked issue: https://github.com/esphome/issues/issues/6540

--- a/components/esp_now/esp_now_component.cpp
+++ b/components/esp_now/esp_now_component.cpp
@@ -86,10 +86,11 @@ void ESPNowComponent::dump_config() { ESP_LOGCONFIG(TAG, "esp_now:"); }
 
 #ifdef USE_ESP8266
 void ESPNowComponent::on_data_received(uint8_t *bssid, uint8_t *data, uint8_t len) {
+  auto packet = make_unique<ESPNowPacket>(bssid, data, len);
 #elif defined(USE_ESP32)
 void ESPNowComponent::on_data_received(const esp_now_recv_info_t *esp_now_info, const uint8_t *data, int len) {
-#endif
   auto packet = make_unique<ESPNowPacket>(esp_now_info->src_addr, data, len);
+#endif
   global_esp_now->receive_queue_.push(std::move(packet));
 }
 

--- a/components/esp_now/esp_now_component.cpp
+++ b/components/esp_now/esp_now_component.cpp
@@ -87,9 +87,9 @@ void ESPNowComponent::dump_config() { ESP_LOGCONFIG(TAG, "esp_now:"); }
 #ifdef USE_ESP8266
 void ESPNowComponent::on_data_received(uint8_t *bssid, uint8_t *data, uint8_t len) {
 #elif defined(USE_ESP32)
-void ESPNowComponent::on_data_received(const uint8_t *bssid, const uint8_t *data, int len) {
+void ESPNowComponent::on_data_received(const esp_now_recv_info_t *esp_now_info, const uint8_t *data, int len) {
 #endif
-  auto packet = make_unique<ESPNowPacket>(bssid, data, len);
+  auto packet = make_unique<ESPNowPacket>(esp_now_info->src_addr, data, len);
   global_esp_now->receive_queue_.push(std::move(packet));
 }
 

--- a/components/esp_now/esp_now_component.h
+++ b/components/esp_now/esp_now_component.h
@@ -1,5 +1,9 @@
 #pragma once
 
+#if defined(USE_ESP32)
+#include <esp_now.h>
+#endif
+
 #include "esphome/core/automation.h"
 #include "esphome/core/component.h"
 #include "esphome/core/helpers.h"
@@ -44,7 +48,7 @@ class ESPNowComponent : public Component {
 #ifdef USE_ESP8266
   static void on_data_received(uint8_t *bssid, uint8_t *data, uint8_t len);
 #elif defined(USE_ESP32)
-  static void on_data_received(const uint8_t *bssid, const uint8_t *data, int len);
+  static void on_data_received(const esp_now_recv_info_t *esp_now_info, const uint8_t *data, int len);
 #endif
 
   Trigger<ESPNowPacket> *get_on_packet_trigger() { return this->on_packet_; }

